### DISCUSSION TITLE
Permissions nesecary to comment on a PR

### DISCUSSION THREAD

**Discussion by WyriHaximus** [community/community#26644] [Actions]
<p>Been working on a workflow that will comment based on changes in the PR. But when the PR comes from the outside or Dependabot it doesn’t have all the write permissions the repo’s owner has. Does anyone know which permissions have to be set to what to make an action be allowed to comment on a PR, when that PR isn’t coming from the repo owner?</p>

**Comment by jsoref** [comment]
<p>Dependabot is fun.</p>
<p>Technically according to the documentation:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/guides/commenting-on-an-issue-when-a-label-is-added#testing-the-workflow" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/guides/commenting-on-an-issue-when-a-label-is-added#testing-the-workflow" target="_blank" rel="noopener nofollow ugc">Commenting on an issue when a label is added - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/guides/commenting-on-an-issue-when-a-label-is-added#testing-the-workflow" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/guides/commenting-on-an-issue-when-a-label-is-added</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>All you need is:</p>
<pre><code class="lang-yaml">    permissions:
      issues: write
</code></pre>
<p>Practically, you need a PAT</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" target="_blank" rel="noopener nofollow ugc">Creating a personal access token - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<p>
And practically, the scopes don’t align well with <code>comment</code>:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/developers/apps/scopes-for-oauth-apps" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/developers/apps/scopes-for-oauth-apps" target="_blank" rel="noopener nofollow ugc">Scopes for OAuth Apps - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/developers/apps/scopes-for-oauth-apps" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/developers/apps/scopes-for-oauth-apps</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>What I’ve managed to do in general is rely on <code>pull_request_target</code><br>
But do read</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/2119212/181099030-e2a91f53-8ac4-49b7-8763-53f53abd9382.png" class="site-icon" width="72" height="72">
      <a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener nofollow ugc">securitylab.github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/2119212/181099031-6b24a590-9d04-497c-bd7a-1a69565510fe.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/e/c/ec60f25ad72688253445273a7394448fe64ea35a_2_10x10.png"></div>

<h3><a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener nofollow ugc">Keeping your GitHub Actions and workflows secure: Preventing pwn requests</a></h3>


  <p><span class="label1">In this article, we’ll discuss some common security malpractices for GitHub Actions and workflows, and how to best avoid them. Our examples are based on real-world GitHub workflow implementation vulnerabilities the GitHub Security Lab has reported to...</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Because if you mess up, the consequences are pretty grave.</p>
<p>It’d be really nice if there was a <code>comment</code> scope available. Because a huge portion of GitHub actins really only need this permission. But, sadly, it doesn’t appear to exist today.</p>

**Reply by RAmi767mlk** [reply]
طيب انا ما عندي فلوس عشان

**Comment by WyriHaximus** [comment]
<p>Thanks! Been really hoping to avoid using a PAT because that widens the attack possibilities with it significantly rather than reducing it. And just setting the <code>issues</code> to <code>write</code> permissions works fine when I make the PR. But any external party still gets blocked, currently attempting the route as described here: <a href="https://github.com/marocchino/sticky-pull-request-comment/issues/298#issuecomment-836278838" class="inline-onebox" rel="noopener nofollow ugc">"Error: Resource not accessible by integration" when using Dependabot · Issue #298 · marocchino/sticky-pull-request-comment · GitHub</a></p>
<p>Still feels as way to much work for something that should be handled more securely by the platform. Right now having to jump through hoops makes it to easy to implement something insecurely.</p>

**Comment by jsoref** [comment]
<p>Yeah, I’m currently playing with variations.</p>
<p>I’m still working on getting all of my pieces to support <code>outputs</code> so that I could talk to that action.</p>
<p>I agree, it’s incredibly inelegant.</p>
<p>Although, one complaint people have about my thing is that it leaves too many comments, whereas a single running comment would be nice, so it’s possible that I’ll want to use that action (I’ve been thinking about it for the past week or two).</p>

**Reply by jsoref** [reply]
I should note that I now have my action automatically collapsing previous comments. 

https://github.com/check-spelling/check-spelling/blob/0dfb89018e1aa8e18e3e108c294d34870161a316/unknown-words.sh

**Comment by WyriHaximus** [comment, marked as answer]
<p>So I currently came to this after a Twitter thread with one of the Dependabot people:</p>
<pre><code class="lang-yaml">name: Composer Diff
on:
  pull_request_target:
    types:
      - opened
      - synchronize
      - reopened
permissions:
  pull-requests: write
jobs:
  comment-composer-lock-diff:
    name: Comment composer.lock diff
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - name: Comment composer.lock diff
        uses: WyriHaximus/github-action-composer.lock-diff@main
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
</code></pre>
<p>For reference that Twitter thread can be found here: <a href="https://twitter.com/WyriHaximus/status/1393679576828686340" rel="noopener nofollow ugc">https://twitter.com/WyriHaximus/status/1393679576828686340</a></p>

**Comment by jsoref** [comment]
<p>Btw, I’m shocked to see <em>anyone</em> using <code>actions/checkout@v1</code> at this point.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/actions/checkout/tree/releases/v1" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h3><a href="https://github.com/actions/checkout/tree/releases/v1" target="_blank" rel="noopener nofollow ugc">actions/checkout</a></h3>

  <p><a href="https://github.com/actions/checkout/tree/releases/v1" target="_blank" rel="noopener nofollow ugc">releases/v1</a></p>

  <p><span class="label1">Action for checking out a repo. Contribute to actions/checkout development by creating an account on GitHub.</span></p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>v1 has been dead since:</p><aside class="onebox githubcommit">
  <header class="source">

      <a href="https://github.com/actions/checkout/commit/0b496e91ec7ae4428c3ed2eeb4c3a40df431f2cc" target="_blank" rel="noopener nofollow ugc">github.com/actions/checkout</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/checkout/commit/0b496e91ec7ae4428c3ed2eeb4c3a40df431f2cc" target="_blank" rel="noopener nofollow ugc">Upgrade checkout to v1.1.0 to better support scripting git. (#56)</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2019-10-25" data-time="14:52:59" data-timezone="UTC">02:52PM - 25 Oct 19 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/TingluoHuang" target="_blank" rel="noopener nofollow ugc">
          <img alt="TingluoHuang" src="https://user-images.githubusercontent.com/2119212/181099034-45ce1c49-c41a-43c6-998b-0f75fcad50fb.png" class="onebox-avatar-inline" width="20" height="20">
          TingluoHuang
        </a>
      </div>

      <div class="lines" title="changed 3 files with 54 additions and 7 deletions">
        <a href="https://github.com/actions/checkout/commit/0b496e91ec7ae4428c3ed2eeb4c3a40df431f2cc" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+54</span>
          <span class="removed">-7</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">* Upgrade checkout to v1.1.0 to better support scripting git.

* add changelog<span class="show-more-container"><a href="https://github.com/actions/checkout/commit/0b496e91ec7ae4428c3ed2eeb4c3a40df431f2cc" target="_blank" rel="noopener nofollow ugc" class="show-more">…</a></span><span class="excerpt hidden">

* Update README.md

* Add local testing yaml

* Add Status Badge to readme</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by WyriHaximus** [comment]
<p>Why? That commit is included in the <code>v1</code> tag, I’m rather shocked that you didn’t notice that as it’s right there when you open the page:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/b/6/b66441054c6b2a46a1ead1d0a24712e3eb2fdda6.png" data-download-href="/uploads/short-url/q1vGX7gZC5ihkuGJBuPPyfViZPU.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/147145/181099036-b9093a62-069c-4642-8f95-011b6f0cf96c.png" alt="image" data-base62-sha1="q1vGX7gZC5ihkuGJBuPPyfViZPU" width="690" height="355" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/b/6/b66441054c6b2a46a1ead1d0a24712e3eb2fdda6_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">811×418 22 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by jsoref** [comment]
<p>Because <code>v2</code> was a major change from <code>v1</code>; <code>v1</code> is clearly not getting updates. So, unless there’s a specific reason to use <code>v1</code>, one really shouldn’t.</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/actions/checkout/releases/tag/v2.0.0" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181099037-f0299115-6f4f-4dc2-9b45-415e7a211523.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/b/ab75a387a76bc1161d73092327b7321c5f1d1f33_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/b/ab75a387a76bc1161d73092327b7321c5f1d1f33_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/b/ab75a387a76bc1161d73092327b7321c5f1d1f33.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/b/ab75a387a76bc1161d73092327b7321c5f1d1f33_2_10x10.png"></div>

<h3><a href="https://github.com/actions/checkout/releases/tag/v2.0.0" target="_blank" rel="noopener nofollow ugc">Release v2.0.0 · actions/checkout</a></h3>

  <p>Improved fetch performance

The default behavior now fetches only the commit being checked-out


Script authenticated git commands

Persists the input token in the local git config
Enables your scr...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by WyriHaximus** [comment]
<p>Ow absolutely, most of my workflows (that I haven’t forgotten to update) are running on <code>v2</code>, especially ones with big repositories. However, that specific action I’ve build has a very specific git use case that I’ve been unable to get working on <code>v2</code>, so yeah it uses <code>v1</code> because that just works.</p>

**Comment by jsoref** [comment]
<p>I should note that while commenting requires permissions, if all you want is to provide a report, there’s a new feature called a job summary:</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/2119212/181099038-796a2a39-9f39-42fe-8d9e-acb41f114733.png" class="site-icon" width="32" height="32">

      <a href="https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/" target="_blank" rel="noopener" title="04:00PM - 09 May 2022">The GitHub Blog – 9 May 22</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181099039-357461f7-37c3-4b13-8ac7-41ce3fe0afab.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/e/3e1fd6f9b5385bcfd41e7cf21146f4eabe01cbf5_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/3/e/3e1fd6f9b5385bcfd41e7cf21146f4eabe01cbf5_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/3/e/3e1fd6f9b5385bcfd41e7cf21146f4eabe01cbf5_2_1380x690.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/e/3e1fd6f9b5385bcfd41e7cf21146f4eabe01cbf5_2_10x10.png"></div>

<h3><a href="https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/" target="_blank" rel="noopener">Supercharging GitHub Actions with Job Summaries | The GitHub Blog</a></h3>

  <p>You can now output and group custom Markdown content on the Actions run summary page.</p>

  <p>
    <span class="label1">Est. reading time: 3 minutes</span>
  </p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>… and it doesn’t require additional permissions.</p>

**Comment by WyriHaximus** [comment]
<p>Does it support updating the same summary if you push again to the same PR?</p>

**Comment by jsoref** [comment]
<p>Summaries are assigned to jobs, so each job gets its own summary.</p>
<p>Roughly it ends up being a non-issue.</p>
<p>Note: I haven’t deployed this feature (for various reasons). I might be able to in June or July.</p>

**Comment by WyriHaximus** [comment]
<p>Also still need to check it out, but if only the latest summary is shown on the PR page it could work. Will see if I can have a try at this soon</p>

**Comment by jsoref** [comment]
<p>I believe they’re basically only reachable via the :white_check_mark:/:x: for the commit or the summary box at the bottom of the PR.</p>

**Comment by sbernard31** [comment]
@WyriHaximus do you have any feedback about this ? (usage of summary OR good way to put comment on PR with github action from forked repo)

I looked at https://securitylab.github.com/research/github-actions-preventing-pwn-requests/ and I ask my self if there is a better way than `pull_request` + `artifact` + `workflow_run`. I feel this is a lot of overhead and would be happy to find something lighter.


**Edit:**

I just see that you propose `pull_request_target` with limited `pull_request:write` permissions but I don't know if in your case it's OK but in my case, I feel an attacker could steal this token to [do some mess on all PRs of my repository](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps#permission-on-pull-requests).  Not sure they could do something very annoying with this access right.

**Reply by WyriHaximus** [reply]
To be honest, the best way forward would be that GitHub provides an "action only permission" where you only can create and update or delete your own comments.

Haven't had the time to look at summaries yet unfortunately.

**Reply by jsoref** [reply]
Stolen tokens are definitely a huge risk, and my initial efforts at workflows indeed were at risk.

I _believe_ my current workflow isn't. But, I could be wrong.

I'd definitely prefer to have a specific permission for `comments` as opposed to having to rely on `pull-requests: write` or `contents: write`. It seems to my naive perspective like a totally reasonable ask.

**Reply by sbernard31** [reply]
>  the best way forward would be that GitHub provides an "action only permission" where you only can create and update or delete your own comments.

Clearly that sounds a really good idea. 
Any way to ping a github dev to share this idea ?

**Reply by jsoref** [reply]
I haven't found an effective channel to do so.

**Reply by sbernard31** [reply]
Reading [this](https://github.com/community/community#making-suggestions), It seems the best way should be to create a discussion and get lot of upvote.

Guys, do you think we should create a new discussion mainly focus on this idea : 

>  GitHub provides an "action only permission" where you only can create and update or delete your own comments.

The idea will be to expose the needs,
then why current solution is not so good,
and finally propose this idea :point_up: 
(of course I put a link to this discussion too)

(I can do that if needed)

Or we should just use this issue and upvote it ?

**Reply by jsoref** [reply]
@WyriHaximus: can you fix the spelling of this one to say `necessary` instead of `nesecary`?

I have no sense of the best way to organize, sorry.
